indexer/bin/piece-indexer.js

```diff
@@ -0,0 +1,58 @@
+import assert from 'assert'
+import { Redis } from 'ioredis'
+import { walkChain } from '../lib/advertisement-walker.js'
+import { runIpniSync } from '../lib/ipni-watcher.js'
+import { RedisRepository } from '../lib/redis-repository.js'
+
+/** @import { ProviderToInfoMap } from '../lib/typings.d.ts' */
+
+const {
+  REDIS_URL: redisUrl = 'redis://localhost:6379'
+} = process.env
+
+// TODO: setup Sentry
+
+const redisUrlParsed = new URL(redisUrl)
+const redis = new Redis({
+  host: redisUrlParsed.hostname,
+  port: Number(redisUrlParsed.port),
+  username: redisUrlParsed.username,
+  password: redisUrlParsed.password,
+  lazyConnect: true, // call connect() explicitly so that we can exit on connection error
+  family: 6 // required for upstash
+})
+
+await redis.connect()
+const repository = new RedisRepository(redis)
+
+/** @type {Set<string>} */
+const providerIdsActivelyWalked = new Set()
+
+/** @type {ProviderToInfoMap} */
+const recentProvidersInfo = new Map()
+
+/**
+ * @param {string} providerId
+ */
+const getProviderInfo = async (providerId) => {
+  const info = recentProvidersInfo.get(providerId)
+  assert(!!info, `Unknown providerId ${providerId}`)
+  return info
+}
+
+for await (const providerInfos of runIpniSync({ minSyncIntervalInMs: 60_000 })) {
+  for (const [providerId, providerInfo] of providerInfos.entries()) {
+    recentProvidersInfo.set(providerId, providerInfo)
+    if (providerIdsActivelyWalked.has(providerId)) continue
+
+    providerIdsActivelyWalked.add(providerId)
+    walkChain({
+      repository,
+      providerId,
+      getProviderInfo,
+      minStepIntervalInMs: 100
+    }).finally(
```

Comment:
Are you worried about the maximum concurrency this can run in?

Reply:
I'm happy to let it fail, debug and then see that maximum concurrency is the issue. It would be something else if we knew yes 100% it is, but like this I think it's fine.

What we need monitoring for is whether the indexer is delayed, I believe as long as it's not, no other metric is important.